src/main/java/org/jetbrains/bsp/bazel/server/DiagnosticsDispatcher.java

```diff
@@ -0,0 +1,124 @@
+package org.jetbrains.bsp.bazel.server;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.Diagnostic;
+import ch.epfl.scala.bsp4j.DiagnosticSeverity;
+import ch.epfl.scala.bsp4j.Position;
+import ch.epfl.scala.bsp4j.PublishDiagnosticsParams;
+import ch.epfl.scala.bsp4j.Range;
+import ch.epfl.scala.bsp4j.SourceItem;
+import ch.epfl.scala.bsp4j.TextDocumentIdentifier;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import io.bazel.rules_scala.diagnostics.Diagnostics;
+import io.bazel.rules_scala.diagnostics.Diagnostics.FileDiagnostics;
+import io.bazel.rules_scala.diagnostics.Diagnostics.Severity;
+import io.bazel.rules_scala.diagnostics.Diagnostics.TargetDiagnostics;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.bsp.bazel.common.Uri;
+
+public class DiagnosticsDispatcher {
+
+  private static final Logger LOGGER = LogManager.getLogger(DiagnosticsDispatcher.class);
+
+  private final Map<Severity, DiagnosticSeverity> CONVERTED_SEVERITY =
+      new ImmutableMap.Builder<Severity, DiagnosticSeverity>()
+          .put(Severity.UNKNOWN, DiagnosticSeverity.ERROR)
+          .put(Severity.ERROR, DiagnosticSeverity.ERROR)
+          .put(Severity.WARNING, DiagnosticSeverity.WARNING)
+          .put(Severity.INFORMATION, DiagnosticSeverity.INFORMATION)
+          .put(Severity.HINT, DiagnosticSeverity.HINT)
+          .build();
+
+  private final BazelBspServer bspServer;
+  private final BuildClient bspClient;
+
+  public DiagnosticsDispatcher(BazelBspServer bspServer, BuildClient bspClient) {
+    this.bspServer = bspServer;
+    this.bspClient = bspClient;
+  }
+
+  public Map<Uri, List<PublishDiagnosticsParams>> collectDiagnostics(
+      BuildTargetIdentifier target, String diagnosticsLocation) {
+    try {
+      TargetDiagnostics targetDiagnostics =
+          TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
+
+      return targetDiagnostics.getDiagnosticsList().stream()
+          .peek(diagnostics -> LOGGER.info("Collected diagnostics at: {}", diagnostics.getPath()))
+          .collect(
+              Collectors.toMap(
+                  diagnostics -> getUriForPath(diagnostics.getPath()),
+                  diagnostics -> convertDiagnostics(target, diagnostics)));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
```

Comment:
Can be streamified

Reply:
Done, but I'm not sure if it can be done without this big forEach function, because we must publish everything from filesToDiagnostics each time directly after adding each source, right?

Reply:
I don't think it can be better